### PR TITLE
Add flow types to render.js

### PR DESCRIPTION
**Pull request by somewhatabstract**
This adds flow types to `render.js` as well as refactoring some code to tidy up how we do things.

Namely:

- Change `configureApolloClient` so that it doesn't need to know how it all gets used
- Add libdef for `argparse`
- Add libdef for `apollo-cache-inmemory`
- Add more strong types to things where we can and increase flow coverage

Before this change:

```shell
npx flow batch-coverage src

Coverage results from 11 file(s):

/Users/jeffyates/khan/react-render-server/src/types.js: 100.00% (18 of 18 expressions)
/Users/jeffyates/khan/react-render-server/src/server.js: 88.17% (395 of 448 expressions)
/Users/jeffyates/khan/react-render-server/src/secret_test.js: 77.01% (134 of 174 expressions)
/Users/jeffyates/khan/react-render-server/src/secret.js: 100.00% (93 of 93 expressions)
/Users/jeffyates/khan/react-render-server/src/profile.js: 100.00% (38 of 38 expressions)
/Users/jeffyates/khan/react-render-server/src/main.js: 84.48% (98 of 116 expressions)
/Users/jeffyates/khan/react-render-server/src/logging.js: 86.89% (179 of 206 expressions)
/Users/jeffyates/khan/react-render-server/src/fetch_package.js: 78.23% (97 of 124 expressions)
/Users/jeffyates/khan/react-render-server/src/create-render-context.js: 86.29% (321 of 372 expressions)
/Users/jeffyates/khan/react-render-server/src/configure-apollo-network.js: 90.11% (82 of 91 expressions)
/Users/jeffyates/khan/react-render-server/src/arguments.js: 56.79% (46 of 81 expressions)

-----------------------------------
Aggregate coverage statistics
-----------------------------------
Files                : 11
Expressions          :
  Covered            : 1501
  Total              : 1761
  Covered Percentage : 85.24%
```

After this change:

```shell
npx flow batch-coverage src

Coverage results from 12 file(s):

/Users/jeffyates/khan/react-render-server/src/types.js: 100.00% (17 of 17 expressions)
/Users/jeffyates/khan/react-render-server/src/server.js: 90.42% (406 of 449 expressions)
/Users/jeffyates/khan/react-render-server/src/secret_test.js: 77.01% (134 of 174 expressions)
/Users/jeffyates/khan/react-render-server/src/secret.js: 100.00% (93 of 93 expressions)
/Users/jeffyates/khan/react-render-server/src/render.js: 83.82% (145 of 173 expressions)
/Users/jeffyates/khan/react-render-server/src/profile.js: 100.00% (38 of 38 expressions)
/Users/jeffyates/khan/react-render-server/src/main.js: 84.48% (98 of 116 expressions)
/Users/jeffyates/khan/react-render-server/src/logging.js: 86.89% (179 of 206 expressions)
/Users/jeffyates/khan/react-render-server/src/fetch_package.js: 78.23% (97 of 124 expressions)
/Users/jeffyates/khan/react-render-server/src/create-render-context.js: 87.10% (324 of 372 expressions)
/Users/jeffyates/khan/react-render-server/src/configure-apollo-network.js: 96.88% (93 of 96 expressions)
/Users/jeffyates/khan/react-render-server/src/arguments.js: 97.65% (83 of 85 expressions)

-----------------------------------
Aggregate coverage statistics
-----------------------------------
Files                : 12
Expressions          :
  Covered            : 1707
  Total              : 1943
  Covered Percentage : 87.85%
```

### CODE REVIEW

File: src/configure-apollo-network.js [hunk 1]
```diff
@@ -10,12 +10,27 @@
  * Requests will automatically timeout after 1000ms, unless another
  * timeout is provided via a 'timeout' property.
  */
-import * as ApolloClient from "apollo-client";
+import * as ApolloClientModule from "apollo-client";
```
Comment (jeresig) [inline src/configure-apollo-network.js]: Oof - that's kind of annoying!

File: src/configure-apollo-network.js [hunk 2]
```diff
@@ -10,12 +10,27 @@
...
 import {InMemoryCache} from "apollo-cache-inmemory";
 import {createHttpLink} from "apollo-link-http";
 import fetch from "node-fetch";
 
-import type {RenderContext} from "./types.js";
+import type {NormalizedCacheObject} from "apollo-cache-inmemory";
+import type {ApolloClient, ApolloCache, ApolloLink} from "apollo-client";
```
Comment (jeresig) [inline src/configure-apollo-network.js]: Alternatively you could do `ApolloClient as ApolloClientType` or some such.

File: src/configure-apollo-network.js [hunk 3]
```diff
@@ -56,26 +72,35 @@ export default function configureApolloNetwork(context: RenderContext): void {
...
-            // HACK(briang): If you give the uri undefined, it will call
-            // fetch("/graphql") but we want to ensure that an undefined URL
-            // will fail the request.
-            uri: ApolloNetwork.url || BAD_URL,
-            fetch: handleNetworkFetch,
-            headers: ApolloNetwork.headers,
-        }),
+        // NOTE(somewhatabstract): We have to cast to any here since the type
+        // of ApolloLink exported from apollo-link-http and the one exported
+        // from apollo-client aren't see as the same type by flow (annoying).
```
Comment (jeresig) [inline src/configure-apollo-network.js]: *seen

...huh that's surprising!

File: src/render.js [hunk 4]
```diff
@@ -115,7 +182,15 @@ const render = async function(jsPackages, props, globals, requestStats) {
 
     try {
         // If Apollo is required, get it configured on the context.
-        configureApolloNetwork(context.window);
+        const apolloNetwork: ?ApolloNetworkConfiguration = (context.window
+            .ApolloNetwork: any);
+        const apolloGlobals = configureApolloNetwork(apolloNetwork);
+
+        /**
```
Comment (jeresig) [inline src/render.js]: Is this new? I can't find where this moved from.

Comment (somewhatabstract) [reply src/render.js]: It was inside the configure method before. Nothing new

**Comment by somewhatabstract** [comment]
> it's interesting that mobile shoots for 100% coverage, I wonder if we could get close with this code base?

I've tried, but it's surprisingly hard. There are a lot of things that just won't get a type, no matter what I try to do.

File: src/configure-apollo-network.js (hunk 1 above)
Comment (somewhatabstract) [reply src/configure-apollo-network.js]: We need the Node version of the module to run inside the browser context, so we pass the whole module in this way. Seemed clearer to have this be named as the module than have it clash with the name of the type within the module too.

**Pull request merged**